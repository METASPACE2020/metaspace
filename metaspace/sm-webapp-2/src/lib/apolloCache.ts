import { defaultDataIdFromObject, InMemoryCache } from '@apollo/client/core'

const nonNormalizableTypes: any[] = ['User', 'DatasetUser', 'DatasetGroup', 'DatasetProject']
export const makeApolloCache = () =>
  new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          dataset: {
            keyArgs: false,
            read(_, { args, toReference }) {
              return toReference({ __typename: 'Dataset', id: args?.id })
            },
          },
          annotation: {
            keyArgs: false,
            read(_, { args, toReference }) {
              return toReference({ __typename: 'Annotation', id: args?.id })
            },
          },
          group: {
            keyArgs: false,
            read(_, { args, toReference }) {
              return toReference({ __typename: 'Group', id: args?.groupId })
            },
          },
          project: {
            keyArgs: false,
            read(_, { args, toReference }) {
              return toReference({ __typename: 'Project', id: args?.projectId })
            },
          },
        },
      },
    },
    dataIdFromObject(object) {
      // WORKAROUND: Because of Apollo's aggressive caching, often the current User will be overwritten with results
      // from other queries. The server side often strips fields based on how they're accessed (the "ScopeRole" logic),
      // which means these query paths will often return different data with the same IDs:
      // currentUser -> primaryGroup (always present)
      // dataset -> submitter -> primaryGroup (null unless admin)
      // To protect against this, don't allow Users (and possibly other types in the future) to have a dataId,
      // so that InMemoryCache cannot share data between different queries.
      if (nonNormalizableTypes.includes(object.__typename)) {
        return false
      } else {
        return defaultDataIdFromObject(object)
      }
    },
  })
