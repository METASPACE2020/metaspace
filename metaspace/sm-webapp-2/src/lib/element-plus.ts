// Importing Element Plus components
import {
  ElPagination,
  ElDialog,
  ElAutocomplete,
  ElDropdown,
  ElDropdownMenu,
  ElDropdownItem,
  ElInput,
  ElInputNumber,
  ElRadio,
  ElRadioGroup,
  ElRadioButton,
  ElCheckbox,
  ElCheckboxButton,
  ElCheckboxGroup,
  ElSwitch,
  ElSelect,
  ElOption,
  ElButton,
  ElButtonGroup,
  ElTable,
  ElTableColumn,
  ElPopover,
  ElTooltip,
  ElForm,
  ElFormItem,
  ElTabs,
  ElTabPane,
  ElTree,
  ElAlert,
  ElSlider,
  ElIcon,
  ElRow,
  ElCol,
  ElProgress,
  ElCarousel,
  ElCarouselItem,
  ElCollapse,
  ElCollapseItem,
  ElDivider,
  ElMessageBox,
  ElMessage,
  ElNotification,
  ElLoading,
  ElOptionGroup,
  ElBadge,
  ElCollapseTransition,
} from 'element-plus'

// Importing Element Plus styles
import 'element-plus/dist/index.css'

const components = {
  ElPagination,
  ElDialog,
  ElAutocomplete,
  ElDropdown,
  ElDropdownMenu,
  ElDropdownItem,
  ElInput,
  ElInputNumber,
  ElRadio,
  ElRadioGroup,
  ElRadioButton,
  ElCheckbox,
  ElCheckboxButton,
  ElCheckboxGroup,
  ElSwitch,
  ElSelect,
  ElOption,
  ElButton,
  ElButtonGroup,
  ElTable,
  ElTableColumn,
  ElPopover,
  ElTooltip,
  ElForm,
  ElFormItem,
  ElTabs,
  ElTabPane,
  ElTree,
  ElAlert,
  ElSlider,
  ElIcon,
  ElRow,
  ElCol,
  ElProgress,
  ElCarousel,
  ElCarouselItem,
  ElCollapse,
  ElCollapseItem,
  ElDivider,
  ElMessage,
  ElNotification,
  ElLoading,
  ElOptionGroup,
  ElBadge,
  ElCollapseTransition,
  // Additional components as needed
}

export {
  ElPagination,
  ElDialog,
  ElAutocomplete,
  ElDropdown,
  ElDropdownMenu,
  ElDropdownItem,
  ElInput,
  ElInputNumber,
  ElRadio,
  ElRadioGroup,
  ElRadioButton,
  ElCheckbox,
  ElCheckboxButton,
  ElCheckboxGroup,
  ElSwitch,
  ElSelect,
  ElOption,
  ElButton,
  ElButtonGroup,
  ElTable,
  ElTableColumn,
  ElPopover,
  ElTooltip,
  ElForm,
  ElFormItem,
  ElTabs,
  ElTabPane,
  ElTree,
  ElAlert,
  ElSlider,
  ElIcon,
  ElRow,
  ElCol,
  ElProgress,
  ElCarousel,
  ElCarouselItem,
  ElCollapse,
  ElCollapseItem,
  ElDivider,
  ElMessage,
  ElMessageBox,
  ElNotification,
  ElLoading,
  ElOptionGroup,
  ElBadge,
  ElCollapseTransition,
  // Additional components as needed
}

export default {
  install(app) {
    Object.values(components).forEach((component: any) => {
      app.component(component.name, component)
    })
    app.config.globalProperties.$message = ElMessage
    app.config.globalProperties.$msgbox = ElMessageBox
    app.config.globalProperties.$alert = ElMessageBox.alert
    app.config.globalProperties.$confirm = ElMessageBox.confirm
    app.config.globalProperties.$prompt = ElMessageBox.prompt
    app.config.globalProperties.$notify = ElNotification
    app.config.globalProperties.$notification = ElNotification
    app.directive('loading', ElLoading.directive)
    app.config.globalProperties.$loading = ElLoading.service
  },
}
