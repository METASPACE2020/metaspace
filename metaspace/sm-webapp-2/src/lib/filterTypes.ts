export interface SingleSelectFilterType {
  value: any
  label: string
  isGroup?: boolean
  id?: any
}

export const datasetOwnerOptions: SingleSelectFilterType[] = [
  { value: '', label: 'All datasets' },
  { value: 'my-datasets', label: 'My datasets' },
]
