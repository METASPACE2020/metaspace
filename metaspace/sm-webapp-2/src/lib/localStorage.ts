import safeJsonParse from './safeJsonParse'
import { useCookies } from 'vue3-cookies'

const memoryStorage: Record<string, string> = {}
const { cookies } = useCookies()
const DAY = 24 * 60 * 60

export const setLocalStorage = (key: string, value: any, cookieFallback = false) => {
  const json = JSON.stringify(value)
  memoryStorage[key] = json
  try {
    localStorage.setItem(key, json)
  } catch (err) {
    console.error(err)
    if (cookieFallback) {
      try {
        cookies.set('storage_' + key, json, 90 * DAY)
      } catch (err2) {
        console.error(err2)
      }
    }
  }
}

export const getLocalStorage = <T>(key: string): T | undefined => {
  try {
    const json = localStorage.getItem(key) || memoryStorage[key] || JSON.stringify(cookies.get('storage_' + key))
    return json && safeJsonParse(json)
  } catch (err) {
    console.error(err)
    return undefined
  }
}

export const removeLocalStorage = (key: string) => {
  delete memoryStorage[key]
  try {
    localStorage.removeItem(key)
  } catch (err) {
    console.error(err)
  }
  try {
    cookies.remove('storage_' + key)
  } catch (err) {
    console.error(err)
  }
}

export const migrateLocalStorage = () => {
  try {
    // Clean up old data stored by Upload page
    removeLocalStorage('latestMetadataSubmission')
    removeLocalStorage('latestMetadataOptions')
    removeLocalStorage('latestMetadataSubmissionVersion')

    // Convert storage cookies to localStorage
    try {
      const hideReleaseNotes = cookies.get('hideReleaseNotes')
      if (hideReleaseNotes != null) {
        setLocalStorage('hideReleaseNotes', hideReleaseNotes, true)
        cookies.remove('hideReleaseNotes')
      }
    } catch (err) {
      console.error(err)
    }
    try {
      const showDescrHint = cookies.get('show_descr_hint') as any
      if (showDescrHint != null) {
        setLocalStorage('hideMarkdownHint', showDescrHint === 0, true)
        cookies.remove('show_descr_hint')
      }
    } catch (err) {
      console.error(err)
    }
  } catch (err) {
    console.error(err)
  }
}
