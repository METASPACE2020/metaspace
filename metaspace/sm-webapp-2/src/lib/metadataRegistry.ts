import metadataMapping from './metadataSchemas/metadataMapping';

export const metadataTypes = Object.keys(metadataMapping);

export const defaultMetadataType = metadataTypes.includes('Imaging MS') ? 'Imaging MS' : metadataTypes[0];

export const metadataSchemas: Record<string, any> = {};

async function loadMetadataSchemas() {
  for (const mdType of metadataTypes) {
    const mdFilename = metadataMapping[mdType];
    try {
      const response = await fetch(`./metadataSchemas/${mdFilename}`);
      if (!response.ok) {
        throw new Error(`Failed to load ${mdFilename}: ${response.statusText}`);
      }
      metadataSchemas[mdType] = await response.json();
    } catch (error) {
      // pass
      // console.error(`Error loading metadata schema for ${mdType}:`, error);
    }
  }
}

loadMetadataSchemas();

export default metadataSchemas;
