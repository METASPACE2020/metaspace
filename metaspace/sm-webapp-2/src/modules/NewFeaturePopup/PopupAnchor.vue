<template>
  <span ref="reference" v-bind="$attrs">
    <slot />
  </span>
</template>
<script lang="ts">
import { defineComponent, computed, ref, watch, onUnmounted } from 'vue'
import Popper, { Placement } from 'popper.js'

import useNewFeaturePopups from './useNewFeaturePopups'
import config from '../../lib/config'
import useIntersectionObserver from '../../lib/useIntersectionObserver'
import { useStore } from 'vuex'

interface Props {
  featureKey: string
  placement: Placement
  showUntil: Date
}

export default defineComponent({
  props: {
    featureKey: String,
    placement: String,
    showUntil: Date,
  },
  setup(props: Props | any) {
    const store = useStore()
    const { activePopup, isDismissed, queuePopup, unqueuePopup, popoverRef } = useNewFeaturePopups()

    const reference = ref<HTMLElement>()

    const isRelevant = (props.showUntil || Number.MAX_VALUE) > new Date()

    if (!isRelevant || isDismissed(props.featureKey)) {
      return {}
    }

    const { isFullyInView } = useIntersectionObserver(reference, { threshold: [0, 1] })

    watch(isFullyInView, (value) => {
      if (value) {
        queuePopup(props.featureKey)
      } else if (activePopup.value !== props.featureKey) {
        // do not unqueue if popup is visible
        unqueuePopup(props.featureKey)
      }
    })

    const isActive = computed(() => {
      return (
        reference.value &&
        popoverRef.value &&
        config.features.new_feature_popups &&
        store.state.currentTour === null &&
        activePopup.value === props.featureKey
      )
    })

    let popper: Popper | null

    watch(isActive, (value) => {
      if (value === true) {
        popper = new Popper(reference.value!, popoverRef.value!, {
          placement: props.placement,
          modifiers: {
            preventOverflow: {
              boundariesElement: 'viewport',
            },
          },
        })
      } else if (popper) {
        popper.destroy()
        popper = null
      }
    })

    onUnmounted(() => {
      if (popper) {
        popper.destroy()
      }
      // in case popup was not dismissed
      unqueuePopup(props.featureKey)
    })

    return {
      reference,
    }
  },
})
</script>
