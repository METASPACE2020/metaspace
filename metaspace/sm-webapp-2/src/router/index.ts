import { createRouter, createWebHistory } from 'vue-router'
import { Component } from 'vue'
import NotFound from '../modules/App/NotFoundPage.vue'
import AboutPage from '../modules/App/AboutPage'
import DatasetsPage from '../modules/Datasets/DatasetsPage.vue'

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ '../modules/Annotations/AnnotationsPage.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ '../modules/MetadataEditor/MetadataEditPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ '../modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/imzml/DatasetBrowserPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/HelpPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Project/ProjectsListPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/PrivacyPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Project/ViewProjectPage.vue'),

}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, Component>

export const routes : any =[
  { path: '/', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/about', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
  {
    path: '/datasets',
    component: DatasetsPage,
    children: [
      { path: '', component: asyncPages.DatasetTable },
      { path: 'summary', component: asyncPages.DatasetTable },
    ],
  },
  {
    path: '/datasets/:dataset_id/comparison',
    name: 'datasets-comparison',
    component: asyncPages.DatasetComparisonPage,
  },
  { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
  { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.DatasetTable },
  { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
  { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
  { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
  { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetTable },
  { path: '/upload', component: asyncPages.UploadPage },
  { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },



  { path: '/groups', component: asyncPages.GroupsListPage },
  { path: '/group/create', component: asyncPages.CreateGroupPage },
  { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },


  { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
  { // Legacy URL sent in "request access" emails up until Feb 2019
    path: '/project/:projectIdOrSlug/manage',
    redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } },
  },
  { path: '/projects', component: asyncPages.ProjectsListPage },



  { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },


  {
    path: '/:pathMatch(.*)*',
    component: NotFound,
  },
]

const router = createRouter({
  // @ts-ignore
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
  scrollBehavior(to, from, savedPosition) {
    // return desired position
    if (savedPosition) {
      return savedPosition;
    }
    return { top: 0 };
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})


const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
