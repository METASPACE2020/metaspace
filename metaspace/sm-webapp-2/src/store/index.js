import { createStore } from 'vuex'
import mutations from './mutations'
import getters from './getters'
import actions from './actions.js'

import { accountModule } from '../modules/Account'

const store = createStore({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    reportError: false,

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: { visible: false },

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE',
    route: {
      path: '',
      params: {},
      query: {},
    },
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule,
  },
})

export default store
