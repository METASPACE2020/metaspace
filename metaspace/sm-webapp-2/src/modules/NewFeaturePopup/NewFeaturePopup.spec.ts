import { mount } from '@vue/test-utils'
import { nextTick } from 'vue'
import NewFeaturePopup from './NewFeaturePopup.vue'
import PopupAnchor from './PopupAnchor.vue'
// import useNewFeaturePopups from './useNewFeaturePopups'
import router from '../../router'
import store from '../../store'

import { ref } from 'vue'
import * as useIntersectionObserver from '../../lib/useIntersectionObserver'
// import * as localStorage from '../../lib/localStorage'

vi.mock('popper.js')
vi.mock('../../lib/useIntersectionObserver')
const mockUseIntersectionObserver: any = useIntersectionObserver
vi.mock('../../lib/localStorage')
// const mockLocalStorage : any = localStorage

function mockIntersectionObserver({ isIntersecting = false, isFullyInView = false, intersectionRatio = 0 }) {
  mockUseIntersectionObserver.default.mockImplementation(() => ({
    isIntersecting: ref(isIntersecting),
    isFullyInView: ref(isFullyInView),
    intersectionRatio: ref(intersectionRatio),
    observe: vi.fn(),
    unobserve: vi.fn(),
  }))
}

describe('NewFeaturePopup', () => {
  const TestNewFeaturePopup = {
    components: {
      PopupAnchor, // full name is mocked
      NewFeaturePopup,
    },
    props: ['showUntil'],
    template: `
      <div>
        <PopupAnchor feature-key="test" :show-until="showUntil">
          <span>Test anchor</span>
        </PopupAnchor>
        <NewFeaturePopup ref="popup" title="Test title" feature-key="test">
          <p>test content</p>
        </NewFeaturePopup>
      </div>
    `,
  }

  beforeEach(() => {
    vi.resetAllMocks()
  })

  it('should match snapshot (not in view)', async () => {
    mockIntersectionObserver({ isFullyInView: false })
    const wrapper = mount(TestNewFeaturePopup, { global: { plugins: [store, router] } })
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should match snapshot (in view)', async () => {
    mockIntersectionObserver({
      isFullyInView: true,
    })
    const wrapper = mount(TestNewFeaturePopup, { global: { plugins: [store, router] } })
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  // TODO: fix this tests
  // it('should close on remind later', async() => {
  //   mockIntersectionObserver({
  //     isFullyInView: true,
  //   })
  //   const { remindLater } = useNewFeaturePopups()
  //
  //   const wrapper = mount(TestNewFeaturePopup, { global: { plugins: [store, router] } })
  //   await nextTick()
  //
  //   remindLater()
  //   await nextTick()
  //
  //   const popup = wrapper.findComponent(NewFeaturePopup)
  //   expect(popup.html()).toBe('')
  // })
})
