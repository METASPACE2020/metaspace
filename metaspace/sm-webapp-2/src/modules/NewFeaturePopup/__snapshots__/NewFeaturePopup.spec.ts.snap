// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewFeaturePopup > should match snapshot (in view) 1`] = `
"<div><span><span>Test anchor</span></span>
  <!--v-if-->
</div>"
`;

exports[`NewFeaturePopup > should match snapshot (not in view) 1`] = `
"<div><span><span>Test anchor</span></span>
  <!--v-if-->
</div>"
`;
